### PR TITLE
Kafka listener

### PR DESCRIPTION
**Pull request by matdieuhau999**
kafka read write data

### CODE REVIEW

File: src/main/java/org/red5/server/kafka/ClientToKafkaStream.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    }
+
+    @Override
+    public void close() {
+        super.close();
+        kafkaSourceListener.close();
+        removeStreamListener(kafkaSourceListener);
+    }
+
+    @Override
```
Comment (daib) [inline src/main/java/org/red5/server/kafka/ClientToKafkaStream.java]: Why do we need these functions? We should not and do not need to use these function.

File: src/main/java/org/red5/server/kafka/KafkaSourceListener.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.Properties;
+
+import org.red5.server.kafka.KafkaProto.KafkaRTMPMessage;
+
+public class KafkaSourceListener implements IStreamListener {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaSourceListener.class);
+
+    private static final String bootstrapServer = "localhost:9092";
```
Comment (daib) [inline src/main/java/org/red5/server/kafka/KafkaSourceListener.java]: The address of bootstrap server should not be hard coded. In the ClientToKafkaStream has kafkaBrokerAddress. Please use the value.

File: src/main/java/org/red5/server/kafka/MyKafkaConsumer.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package org.red5.server.kafka;
```
Comment (daib) [inline src/main/java/org/red5/server/kafka/MyKafkaConsumer.java]: This class should not be checked in.

File: src/main/java/org/red5/server/stream/ClientBroadcastStream.java [hunk 4]
```diff
@@ -231,6 +231,7 @@ public void close() {
      *            Event to dispatch
      */
     public void dispatchEvent(IEvent event) {
+        log.info("this is dispatchEvent ClientBroadcastStream");
```
Comment (daib) [inline src/main/java/org/red5/server/stream/ClientBroadcastStream.java]: Remove this. Should not change the original code.

File: src/main/java/org/red5/server/stream/ClientBroadcastStream.java [hunk 5]
```diff
@@ -586,6 +587,8 @@ public void onOOBControlMessage(IMessageComponent source, IPipe pipe, OOBControl
      */
     @SuppressWarnings("unused")
     public void onPipeConnectionEvent(PipeConnectionEvent event) {
+        log.info("onPipeConnectionEvent");
```
Comment (daib) [inline src/main/java/org/red5/server/stream/ClientBroadcastStream.java]: Remove this line.

File: replay_pid9504.log [hunk 6]
```diff
@@ -0,0 +1,4 @@
+JvmtiExport can_access_local_variables 0
+JvmtiExport can_hotswap_or_post_breakpoint 0
+JvmtiExport can_post_on_exceptions 0
+# 304 ciObject found
```
Comment (daib) [inline replay_pid9504.log]: Remove this file from commit.

File: src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Override
+    public void onPipeConnectionEvent(PipeConnectionEvent event) {
+        super.onPipeConnectionEvent(event);
+        myLivePipe = livePipe;
+        livePipe = null;
+    }
+
+    //KafkaThread class
```
Comment (daib) [inline src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java]: This comment does not mean anything ...

File: src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+ * This class reads from Kafka and broadcasts to subscribers
+ */
+@ManagedResource(objectName = "org.red5.server:type=KafkaStreamBroadcaster", description = "KafkaStreamBroadcaster")
+public class KafkaStreamBroadcaster extends ClientBroadcastStream {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaStreamBroadcaster.class);
+
+    protected String kafkaBrokerAddress;
+
+    private static final String bootstrapServer = "localhost:9092";
```
[the remark this answers is not in the document]
Comment (matdieuhau999) [reply src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java]: I forgot removimg it. It isn't used

File: src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+                        }
+                    });
+                } else {
+                    log.debug("Live pipe was null, message was not pushed");
+                }
+            }
+        }
+    }
+
+    @Override
```
[the remark this answers is not in the document]
Comment (matdieuhau999) [reply src/main/java/org/red5/server/kafka/KafkaStreamBroadcaster.java]: I have remove all. It is not the last commit. In my recent KafkaStreamBroadcaster file have128 line

File: src/main/java/org/red5/server/kafka/KafkaSourceListener.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+    private Properties propertiesProducer;
+
+    private KafkaProducer<String, byte[]> kafkaProducer;
+
+    private static final String TOPIC = "test";
```
[the remark this answers is not in the document]
Comment (matdieuhau999) [reply src/main/java/org/red5/server/kafka/KafkaSourceListener.java]: I have removed it. It is not the last commit

File: replay_pid9504.log (hunk 6 above)
Comment (matdieuhau999) [reply replay_pid9504.log]: it is removed

File: src/main/java/org/red5/server/kafka/MyKafkaConsumer.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.red5.server.kafka.KafkaProto.KafkaRTMPMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Properties;
+
+public class MyKafkaConsumer {
```
Comment (daib) [inline src/main/java/org/red5/server/kafka/MyKafkaConsumer.java]: Rename this class

**Pull request merged**